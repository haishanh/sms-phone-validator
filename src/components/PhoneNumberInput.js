import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import isPhoneNumberValid from '../helpers/validatePhoneNumber';

class PhoneNumberInput extends Component {
  state = {
    value: '',
    isInputValid: false
  };

  static propTypes = {
    placeholder: PropTypes.string.isRequired,
    onValidationChange: PropTypes.func
  };

  static defaultProps = {
    onValidationChange: () => {}
  };

  handleOnChange = ev => {
    ev.preventDefault();
    const { value } = ev.target;
    const { isInputValid } = this.state;
    const isInputValidNext = isPhoneNumberValid(value);
    this.setState({ value, isInputValid: isInputValidNext });
    if (isInputValid !== isInputValidNext) {
      this.props.onValidationChange(isInputValidNext);
    }
  };

  render() {
    const { placeholder } = this.props;
    const { value } = this.state;
    return (
      <Fragment>
        <input
          type="tel"
          maxLength="11"
          placeholder={placeholder}
          value={value}
          onChange={this.handleOnChange}
        />
      </Fragment>
    );
  }
}

export default PhoneNumberInput;
