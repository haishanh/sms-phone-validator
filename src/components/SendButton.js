import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import sendSms from '../helpers/sendSms';

class SendButton extends Component {
  initialState = {
    remaining: -1,
    tickingTimeoutId: null
  };

  state = this.initialState;

  static defaultProps = {
    sendSmsBackoffTime: 60,
    fetchImpl: window.fetch
  };

  static propTypes = {
    sendSmsBackoffTime: PropTypes.number,
    fetchImpl: PropTypes.func,
    text: PropTypes.string.isRequired,
    canSend: PropTypes.bool.isRequired,
    onCountdownStarted: PropTypes.func.isRequired,
    onCountdownCompleted: PropTypes.func.isRequired,
    onSendFailed: PropTypes.func.isRequired
  };

  ticking(remaining, callback) {
    this.setState({ remaining });
    if (remaining >= 1) {
      const tickingTimeoutId = setTimeout(() => {
        this.ticking(remaining - 1, callback);
        // zhi yao jiu jiu ba
      }, 998);
      this.setState({ tickingTimeoutId });
    } else {
      this.setState({ tickingTimeoutId: null });
      callback();
    }
  }

  stopCountdown = () => {
    clearTimeout(this.state.tickingTimeoutId);
    this.setState(this.initialState);
  };

  handleButtonOnClick = () => {
    const { fetchImpl } = this.props;
    sendSms(fetchImpl).catch(() => {
      this.stopCountdown();
      this.props.onSendFailed();
    });
    const { sendSmsBackoffTime } = this.props;
    this.props.onCountdownStarted();
    this.ticking(sendSmsBackoffTime, this.props.onCountdownCompleted);
  };

  componentWillUnmount() {
    const { tickingTimeoutId } = this.state;
    if (tickingTimeoutId) clearTimeout(tickingTimeoutId);
    this.setState(this.initialState);
  }

  render() {
    const { canSend, text } = this.props;
    const { remaining, tickingTimeoutId } = this.state;
    let buttonText = text;
    if (tickingTimeoutId) {
      // tick in progress
      buttonText = `(${remaining})${text}`;
    }
    return (
      <Fragment>
        <button disabled={!canSend} onClick={this.handleButtonOnClick}>
          {buttonText}
        </button>
      </Fragment>
    );
  }
}

export default SendButton;
