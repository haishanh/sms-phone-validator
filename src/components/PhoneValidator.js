import React, { Component } from 'react';
import PropTypes from 'prop-types';

import PhoneNumberInput from './PhoneNumberInput';
import SendButton from './SendButton';

// function fetchMockGood() {
//   const res = {
//     ok: true,
//     json: () => {}
//   };
//   return new Promise((resolve, reject) => {
//     return resolve(res);
//   });
// }

// function fetchMockBad() {
//   return new Promise((resolve, reject) => {
//     setTimeout(reject, 3000);
//   });
// }

class PhoneValidator extends Component {
  static defaultProps = {
    buttonTextDefault: '获取验证码',
    buttonTextSent: '已发送',
    placeholder: '手机号',
    sendSmsBackoffTime: 60,
    className: 'phone-validator',
    inputName: 'phoneNumber',
    fetchImpl: fetch
  };

  static propTypes = {
    buttonTextDefault: PropTypes.string,
    buttonTextSent: PropTypes.string,
    placeholder: PropTypes.string,
    sendSmsBackoffTime: PropTypes.number,
    fetchImpl: PropTypes.func
  };

  state = {
    isPhoneNumberValid: false,
    isVerifyInProgress: false
  };

  handleOnValidationChange = isPhoneNumberValid => {
    this.setState({ isPhoneNumberValid });
  };

  onCountdownStarted = () => {
    this.setState({
      isVerifyInProgress: true
    });
  };

  onCountdownCompleted = () => {
    this.setState({
      isVerifyInProgress: false
    });
  };

  onSendFailed = () => {
    this.setState({
      isVerifyInProgress: false
    });
  };

  render() {
    const { isPhoneNumberValid, isVerifyInProgress } = this.state;
    const { buttonTextDefault, buttonTextSent, className, inputName } = this.props;
    const canSend = isPhoneNumberValid === true && isVerifyInProgress === false;
    const buttonText = isVerifyInProgress ? buttonTextSent : buttonTextDefault;

    return (
      <div className={className}>
        <PhoneNumberInput
          onValidationChange={this.handleOnValidationChange}
          name={inputName}
          {...this.props}
        />
        <SendButton
          text={buttonText}
          canSend={canSend}
          onCountdownStarted={this.onCountdownStarted}
          onCountdownCompleted={this.onCountdownCompleted}
          onSendFailed={this.onSendFailed}
          {...this.props}
        />
      </div>
    );
  }
}

export default PhoneValidator;
