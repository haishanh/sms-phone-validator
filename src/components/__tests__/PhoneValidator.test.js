import React from 'react';
import PhoneValidator from '../PhoneValidator';
import PhoneNumberInput from '../PhoneNumberInput';
import SendButton from '../SendButton';
import { shallow, mount } from 'enzyme';

function sleep(t) {
  return new Promise(r => setTimeout(r, t));
}
const noop = () => {};
// const node = document.createElement('div')
function fetchMockGood() {
  const res = {
    ok: true,
    json: () => {}
  };
  return new Promise((resolve, reject) => {
    return resolve(res);
  });
}

// local helper
function getButtonText(wrapper) {
  return (
    wrapper
      .find('button')
      .text()
      // to remove "(12)" in text like "(12)sending"
      .replace(/\([\d]+?\)/, '')
  );
}

describe('<PhoneValidator />', () => {
  it('should render', () => {
    const wrapper = shallow(<PhoneValidator />);
    expect(wrapper.find(PhoneNumberInput).length).toBe(1);
    expect(wrapper.find(SendButton).length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state correctly', () => {
    // full mount
    const wrapper = mount(<PhoneValidator fetchImpl={fetchMockGood} />);
    const fakeClickEvent = {
      preventDefault: noop,
      target: {
        value: '188'
      }
    };
    wrapper.find('input').simulate('change', fakeClickEvent);
    expect(wrapper.state()).toEqual({
      isPhoneNumberValid: false,
      isVerifyInProgress: false
    });

    fakeClickEvent.target.value = '18812341234';
    wrapper.find('input').simulate('change', fakeClickEvent);
    expect(wrapper.state()).toEqual({
      isPhoneNumberValid: true,
      isVerifyInProgress: false
    });
    wrapper.find('button').simulate('click');
    expect(wrapper.state()).toEqual({
      isPhoneNumberValid: true,
      isVerifyInProgress: true
    });
  });

  it('should work as full flow', async () => {
    // full mount
    const buttonTextDefault = 'lorem-buttonTextDefault';
    const buttonTextSent = 'lorem-buttonTextSent';
    const fetchImpl = async () => {
      await sleep(10);
      throw Error('lorem');
    };
    const wrapper = mount(
      <PhoneValidator
        fetchImpl={fetchImpl}
        buttonTextDefault={buttonTextDefault}
        buttonTextSent={buttonTextSent}
      />
    );
    const fakeClickEvent = {
      preventDefault: noop,
      target: {
        value: '18812345678'
      }
    };
    expect(wrapper.find('button').props().disabled).toBe(true);
    expect(getButtonText(wrapper)).toBe(buttonTextDefault);
    expect(wrapper.state()).toEqual({
      isPhoneNumberValid: false,
      isVerifyInProgress: false
    });

    wrapper.find('input').simulate('change', fakeClickEvent);
    expect(wrapper.find('button').props().disabled).toBe(false);

    wrapper.find('button').simulate('click');
    expect(getButtonText(wrapper)).toBe(buttonTextSent);
    expect(wrapper.state()).toEqual({
      isPhoneNumberValid: true,
      isVerifyInProgress: true
    });
    await sleep(30);
    expect(wrapper.state()).toEqual({
      isPhoneNumberValid: true,
      isVerifyInProgress: false
    });
    expect(wrapper.find('button').props().disabled).toBe(true);
    expect(getButtonText(wrapper)).toBe(buttonTextDefault);
  });
});
