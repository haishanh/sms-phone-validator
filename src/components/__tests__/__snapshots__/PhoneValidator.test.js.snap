// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneValidator /> should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhoneValidator
    buttonTextDefault="获取验证码"
    buttonTextSent="已发送"
    fetchImpl={[Function]}
    placeholder="手机号"
    sendSmsBackoffTime={60}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <PhoneNumberInput
          buttonTextDefault="获取验证码"
          buttonTextSent="已发送"
          fetchImpl={[Function]}
          onValidationChange={[Function]}
          placeholder="手机号"
          sendSmsBackoffTime={60}
        />,
        <SendButton
          buttonTextDefault="获取验证码"
          buttonTextSent="已发送"
          canSend={false}
          fetchImpl={[Function]}
          onCountdownCompleted={[Function]}
          onCountdownStarted={[Function]}
          onSendFailed={[Function]}
          placeholder="手机号"
          sendSmsBackoffTime={60}
          text="获取验证码"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "buttonTextDefault": "获取验证码",
          "buttonTextSent": "已发送",
          "fetchImpl": [Function],
          "onValidationChange": [Function],
          "placeholder": "手机号",
          "sendSmsBackoffTime": 60,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "buttonTextDefault": "获取验证码",
          "buttonTextSent": "已发送",
          "canSend": false,
          "fetchImpl": [Function],
          "onCountdownCompleted": [Function],
          "onCountdownStarted": [Function],
          "onSendFailed": [Function],
          "placeholder": "手机号",
          "sendSmsBackoffTime": 60,
          "text": "获取验证码",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PhoneNumberInput
            buttonTextDefault="获取验证码"
            buttonTextSent="已发送"
            fetchImpl={[Function]}
            onValidationChange={[Function]}
            placeholder="手机号"
            sendSmsBackoffTime={60}
          />,
          <SendButton
            buttonTextDefault="获取验证码"
            buttonTextSent="已发送"
            canSend={false}
            fetchImpl={[Function]}
            onCountdownCompleted={[Function]}
            onCountdownStarted={[Function]}
            onSendFailed={[Function]}
            placeholder="手机号"
            sendSmsBackoffTime={60}
            text="获取验证码"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "buttonTextDefault": "获取验证码",
            "buttonTextSent": "已发送",
            "fetchImpl": [Function],
            "onValidationChange": [Function],
            "placeholder": "手机号",
            "sendSmsBackoffTime": 60,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "buttonTextDefault": "获取验证码",
            "buttonTextSent": "已发送",
            "canSend": false,
            "fetchImpl": [Function],
            "onCountdownCompleted": [Function],
            "onCountdownStarted": [Function],
            "onSendFailed": [Function],
            "placeholder": "手机号",
            "sendSmsBackoffTime": 60,
            "text": "获取验证码",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
